docs/snapshotting/versioning.md

```diff
@@ -0,0 +1,98 @@
+
+Below is an example graph showing backward and forward snapshot compatibility for a version space formally represented as `[v2; vn]`. This is the general picture, but keep in mind that future corner cases might remove edges from this graph.
+
+![Version graph](
+../images/version_graph.png?raw=true
+"Version graph")
+
+## Overview
+Firecracker persists the microVM state as 2 separate objects:
+ - a **guest memory** file
+ - a **microVM state** file.
+ 
+*The block devices attached to the microVM are not considered part of the state and need to be managed separately.*
+
+### Guest memory
+The guest memory file is agnostic to Firecracker's version tolerant save/restore implementation as its contents are fully owned by the guest OS. The microVM memory is by choice saved as a dump of each page or as a sparse file of guest modified pages only. 
+### MicroVM state
+The microVM state file stores the internal state of the VMM (device emulation, KVM and vCPUs) with 2 exceptions - serial emulation and vsock backend. 
+
+While we continuously improve and extend Firecracker's features by adding new capabilities, devices or enhancements, the microVM state file may change both structurally and semantically with each new release. The state file includes versioning information and each Firecracker release implements distinct save/load logic for the supported version space. 
+
+## MicroVM state file format
+
+###  A closer look
+A microVM state file is further split into four different fields:
+
+| Field | Bits| Description |
+|----|----|----|
+| magic_id | 64 | Firecracker snapshot, architecture (x86_64/aarch64) and storage version.
+| version  | 64 | The snapshot version number internally mapped 1:1 to a specific Firecracker version.
+| state | N*8 | Bincode blob containing the microVM state.
+ | crc| 64 | Optional CRC64 sum of magic_id, version and state fields.
+
+Note: the last 16 bits of `magic_id` encode the storage version which specifies the encoding used for the `version` and `state` fields. The current implementation sets this field to 1, which identifies it as a [Serde bincode](https://github.com/servo/bincode) compatible encoder/decoder.
+
+### Version tolerant ser/de
+Firecracker reads and writes the `state` blob of the snapshot by using per version, separate serialization and deserialization logic. This logic is mostly autogenerated by a Rust procedural macro based on `struct`, `enum` and `union` annotations. Basically, one can say that these structures support versioning. The versioning logic is generated by parsing a structure's history log (encoded using Rust annotations) and emitting Rust code.
+
+Versioned serialization and deserialization is divided into two translation layers:
+ - structural translator,
+ - semantic translator.
+
+The _structural translator_ implements the logic to convert between different versions of the same Rust POD structure: it can deserialize from source version to target or serialize from source to target. 
```

Comment:
```suggestion
The _structural translator_ implements the logic to convert between different versions of the same Rust POD structure: it can deserialize or serialize from source version to target. 
```